socs/agents/hwp_supervisor/agent.py

```diff
@@ -0,0 +1,773 @@
+            self.hwp_state.update_pid_state(pid_op)
+            self.hwp_state.update_pmx_state(pmx_op)
+            self.hwp_state.update_temp_state(temp_op)
+            self.hwp_state.update_ups_state(ups_op)
+            self.hwp_state.update_enc_state(enc_op)
+            session.data['hwp_state'] = asdict(self.hwp_state)
+
+            # Get actions for each hwp subsystem
+            session.data['actions'] = {
+                'pmx': self.hwp_state.pmx_action,
+                'gripper': self.hwp_state.gripper_action,
+            }
+
+            if test_mode:
+                break
+
+            pm.sleep()
+
+        return True, "Monitor process stopped"
+
+    def _stop_monitor(self, session, params):
+        session.status = 'stopping'
+        return True, 'Stopping monitor process'
+
+    def spin_control(self, session, params):
+        """spin_control()
+
+        **Process** - Process to manage the spin-state for HWP agents. This will
+        issue commands to various HWP agents depending on the current control
+        state.
+        """
+        clients = self._get_hwp_clients()
+
+        while session.status in ['starting', 'running']:
+            try:
+                self.control_state.update(clients, self.hwp_state)
+            except Exception:
+                msg = f"Error updating control state:\n{traceback.format_exc()}"
+                self.log.error(msg)
+                self.control_state.set_state_type(ControlStateType.ERROR, msg=msg)
+
+            session.data = {
+                'state': self.control_state.encoded(),
+                'timestamp': time.time(),
+            }
+            time.sleep(1)
+
+    def _stop_spin_control(self, session, params):
+        session.status = 'stopping'
+        return True, 'Stopping spin control process'
+
+    @ocs_agent.param('target_freq', type=float)
+    @ocs_agent.param('freq_thresh', type=float, default=0.05)
+    @ocs_agent.param('freq_thresh_duration', type=float, default=10)
+    def pid_to_freq(self, session, params):
+        """pid_to_freq(target_freq=2.0, freq_thresh=0.05, freq_thresh_duration=10)
+
+        **Task** - Sets the control state to PID the HWP to the given ``target_freq``.
+
+        Args
+        -------
+        target_freq : float
+            Target frequency of the HWP (Hz). Positive values are CW, negative are CCW.
+        freq_thresh : float
+            Frequency threshold (Hz) for determining when the HWP is at the target frequency.
+        freq_thresh_duration : float
+            Duration (seconds) for which the HWP must be within ``freq_thresh`` of the
+            ``target_freq`` to be considered successful.
+        """
+        if params['target_freq'] >= 0:
+            d = '0' if self.forward_is_cw else '1'
+        else:
+            d = '1' if self.forward_is_cw else '1'
+
+        self.control_state = ControlState(
+            state_type=ControlStateType.PID_FREQ,
+            target_freq=params['target_freq'],
```

Comment:
I believe that this should be `target_freq=abs(params['target_freq'])` as the function it calls only accepts positive values